src/libs/actions/OnyxUpdateManager.ts

```diff
@@ -27,6 +28,132 @@ Onyx.connect({
     callback: (value) => (lastUpdateIDAppliedToClient = value),
 });
 
+type DeferredUpdatesDictionary = Record<number, OnyxUpdatesFromServer>;
+let deferredUpdateBeforeReconnect: OnyxUpdatesFromServer | undefined;
+let deferredUpdates: DeferredUpdatesDictionary = {};
+
+// This function will unpause the SequentialQueue, log an info and reset the deferred updates
+function unpauseQueueAndReset() {
+    console.debug('[OnyxUpdateManager] Done applying all updates');
+    deferredUpdates = {};
+    Onyx.set(ONYXKEYS.ONYX_UPDATES_FROM_SERVER, null);
+    SequentialQueue.unpause();
+}
+
+// This function applies a list of updates to Onyx in order and resolves when all updates have been applied
+const applyUpdates = (updates: DeferredUpdatesDictionary) => Promise.all(Object.values(updates).map((update) => OnyxUpdates.apply(update)));
+
+type DetectGapAndSplitResult = {applicableUpdates: DeferredUpdatesDictionary; updatesAfterGaps: DeferredUpdatesDictionary; latestMissingUpdateID: number | undefined};
+
+// In order for the deferred updates to be applied correctly in order,
+// we need to check if there are any gaps deferred updates.
+function detectGapsAndSplit(updates: DeferredUpdatesDictionary): DetectGapAndSplitResult {
+    const updateValues = Object.values(updates);
+    const applicableUpdates: DeferredUpdatesDictionary = {};
+
+    let gapExists = false;
+    let firstUpdateAfterGaps: number | undefined;
+
+    for (const [index, update] of updateValues.entries()) {
+        // If it's the first update, we need to check that the previousUpdateID of the fetched update is the same as the lastUpdateIDAppliedToClient
+        // If any update's previousUpdateID doesn't match the lastUpdateID from the previous update, there's a gap in the deferred updates
+        const isFirst = index === 0;
+        if (isFirst && update.previousUpdateID !== lastUpdateIDAppliedToClient) {
+            break;
+        }
```

Comment:
Does this means that if we have a gap in the first update already, we will stop here and not check if there are bigger gaps?

Reply:
ah yes, you're right. this is wrong